File: multiloessplot.py
def multiloessplot(x, index=0, low=0.1, high=0.9, step=0.01, tick_step=0.1, title=""):
    """

    multiloessplot

    Plot multiple loess smoothers for motion picture shot length data.

    Parameters
    ----------
    x : a data frame in wide format containing shot length data for a motion picture.

    index : the index of the data column in the data frame to be plotted.
            Note that the index is in python format and begins at 0.

    low : the minimum value of the loess smoothers.

    high : the maximum value of the loess smoothers.

    step : the increment in the span of the loess smoothers.

    tick_step : specifies the distance between tick marks on the colour bar in the legend. 
                The lower and upper limits of the colour bar are set by low and high, respectively.

    title : a title for the plot.

    Returns
    -------
    a plot of the multiple loess smoothers, with the cumulative running time as a percentage plotted 
    against the fitted values of the local regression

    Examples
    --------
    # Load a .csv file containing shot length data
    import pandas as pd
    df = pd.read_csv('path/to/file/data.csv', delimiter=',')

    multiloessplot(df, index=0, low=0.1, high=0.9, step=0.01, tick_step = 0.1, title = "Movie title")
    """

    import numpy as np
    import pandas as pd
    import seaborn as sns
    from statsmodels.nonparametric.smoothers_lowess import lowess
    from matplotlib import cm, colorbar, colors

    SL = x.iloc[:, index].dropna(axis=0, how='any')
    times = np.cumsum(SL)
    times = 100 * (times/np.max(times))

    spans = np.round(np.arange(low, high+step, step).tolist(), 2)

    # Collect the results of fitting multiple smoothers
    column_names = ["span", "times", "fit"]
    dat_out = pd.DataFrame(columns=column_names)

    for s in spans:
        fit = pd.Series(lowess(SL, times, is_sorted=True, frac=s, it=0)[:, 1])
        span = pd.Series(np.repeat(s, len(times)))
        dat_res = pd.DataFrame({"span": span, "times": times, "fit": fit})
        dfs = [dat_out, dat_res]
        dat_out = pd.concat(dfs, ignore_index=True)

    # Create the plot
    sns.set_theme(style="whitegrid", font_scale=1.2)
    p = sns.relplot(data=dat_out, x="times", y="fit", hue="span", kind="line",
                    palette="viridis",
                    legend=False, height=6, aspect=1.3)
    p.set(title=title, xlabel="Running time (%)", ylabel="Fitted values (s)")
    p.figure.colorbar(
        cm.ScalarMappable(
            norm=colors.Normalize(vmin=dat_out['span'].min(),
                                  vmax=dat_out['span'].max(),
                                  clip=False), cmap='viridis'),
        ticks=np.arange(low, high+step, tick_step),
        label=r'Span', orientation='horizontal', shrink=0.8, aspect=20, pad=0.15)

File: MultiLoessPlot.jl
"""
    MultiLoessPlot(df; index, low, step, high, title)

Fit multiple loess smoothers to motion picture shot length data and visualise the result.

# Arguments:
- `df`: a data frame containing shot length data in wide format.
- `index::Integer=1`: the index of the data frame column containing the shot length data to be visulaised.
- `low::Float`=0.1: the minimum loess span.
- `step::Float=0.01`: the increment of the span of the loess smoothers.
- `high::Float=0.9`: the maximum loess spans.
- `title::Char`: the title of the plot.

# Example:
```Julia
# Load a .csv file containing shot length data
using CSV
using DataFrames
df = CSV.read("./path/to/file/file.csv", DataFrame; header=1)

# Display the smoothers
using ColorSchemes
using Gadfly
using Loess
p = MultiLoessPlot(df; index=1, low=0.1, step=0.01, high=0.9, title="Movie Title")

# Save as a .png file
using Cairo
using Fontconfig
p |> PNG("./plot.PNG")
```
"""
function MultiLoessPlot(df; index=1, low=0.1, step=0.01, high=0.9, title="")

    # select data and calculate shot times
    dat = dropmissing(select(df, index))
    SL = dat[!, 1]
    times = cumsum(dat[!, 1])
    times = round.(100 * (times / maximum(times)); digits=2)

    # empty data frame to store the result
    dat_out = DataFrame(span=Float64[], times=Float64[], fit=Float64[])

    # iterte over spans of loess smoothers
    spans = [low:step:high;]

    for s in spans

        model = loess(times, SL, span=s)
        fitted = round.(predict(model, times); digits=2)
        dat_res = DataFrame(span=s, times=times, fit=fitted)
        dat_out = vcat(dat_out, dat_res, cols=:union)

    end

    # create the plot
    cpalette(p) = get(ColorSchemes.viridis, p)

    plot(dat_out,
        x=:times, y=:fit, group=:span, color=:span, Geom.line,
        Scale.color_continuous(colormap=cpalette, minvalue=low, maxvalue=high),
        Guide.title(title), Guide.xlabel("Running time (%)"), Guide.ylabel("Fitted values (s)"),
        Guide.colorkey(title="Span"),
        Guide.xticks(ticks=[0:20:100;]),
        Gadfly.Theme(background_color="white", grid_color="#808080", grid_line_style=:solid,
            key_title_font="Arial", major_label_font="Arial", key_position=:right))

end
